### PR TITLE
HL-627

### PR DESCRIPTION
**Pull request by bazmatic**
https://dltx.atlassian.net/browse/HL-627

Was not managing the totalExposure and totalCollateral correctly.
Added test to cover case where some spare collateral is in the market, not allocated to any race.

### CODE REVIEW

File: contracts/Market.sol [hunk 1]
```diff
@@ -13,6 +13,8 @@ import "./IOracle.sol";
 import "./SignatureLib.sol";
 import "./OddsLib.sol";
 
+import "hardhat/console.sol";
```
Comment (brettvniekerk) [inline contracts/Market.sol:16]: do we want to merge this to prod?

File: contracts/Market.sol [hunk 2]
```diff
@@ -196,7 +198,7 @@ contract Market is IMarket, Ownable, ERC721 {
 			return 1;
 		}
 
-		pool -= _potentialPayout[propositionId]; //TODO: Should be _totalExposure;
+		pool -= _potentialPayout[propositionId];
```
Comment (brettvniekerk) [inline contracts/Market.sol:201]: previously left comment said this should be `_totalExposure`, we don't want to do that anymore?

File: contracts/MarketCollateralised.sol [hunk 3]
```diff
@@ -89,21 +106,22 @@ abstract contract MarketCollateralised is Market {
...
 				_totalCollateral += result;
 			}
+			// Add the amount we borrowed to the race collateral
 			_marketCollateral[marketId] += result;
 		}
-
+		// Store the amount of exposure that this bet added. This will be deducted from the total exposure when the bet is paid out
 		_betExposure[_bets.length] = result;
-		return result;
+		return result; // To be added to total exposure
```
Comment (brettvniekerk) [inline contracts/MarketCollateralised.sol:119]: just double checking that it _does_ get added and the comment isn't a TODO?

File: scripts/historicalBalances.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		return balance;
+	} catch (error) {
+		console.log(error);
+		return 0;
+	}
+}
+
+// We recommend this pattern to be able to use async/await everywhere
+// and properly handle errors.
+main().catch((error) => {
```
Comment (brettvniekerk) [inline scripts/historicalBalances.ts:105]: no exit process?

File: contracts/Market.sol (hunk 2 above)
Comment (bazmatic) [reply contracts/Market.sol:199]: correct

File: contracts/MarketCollateralised.sol (hunk 3 above)
Comment (bazmatic) [reply contracts/MarketCollateralised.sol:118]: It is just a tip, letting you know what will happen with the result. Sounds like a TODO so I'll change the wording to make it clearer

File: scripts/historicalBalances.ts (hunk 4 above)
Comment (bazmatic) [reply scripts/historicalBalances.ts:105]: Oh this was just a little tool I made but didn't intend to push

File: contracts/MarketCollateralised.sol [hunk 5]
```diff
@@ -68,19 +69,34 @@ abstract contract MarketCollateralised is Market {
 		bytes16 propositionId
 	) internal virtual returns (uint256) {
 		uint256 result;
+
 		uint256 internalCollateralToUse;
```
Comment (bitcoinbrisbane) [inline contracts/MarketCollateralised.sol:73]: Can we move this var creation to LN: 79?

File: tests/market.collateralised.test.ts [hunk 6]
```diff
@@ -23,9 +23,10 @@ import {
 	TestMarket
 } from "./utils";
 import { formatBytes16String } from "../scripts/utils";
+import * as timeHelper from "@nomicfoundation/hardhat-network-helpers/dist/src/helpers/time";
 
 chai.use(solidity);
-describe("Collateralised Market: play through", function () {
+describe.only("Collateralised Market: play through", function () {
```
Comment (bitcoinbrisbane) [inline tests/market.collateralised.test.ts:29]: Oops

File: contracts/MarketCollateralised.sol (hunk 5 above)
Comment (bazmatic) [reply contracts/MarketCollateralised.sol:73]: I don't follow

**Review by brettvniekerk — APPROVED** [review]

File: contracts/MarketCollateralised.sol (hunk 5 above)
Comment (bitcoinbrisbane) [reply contracts/MarketCollateralised.sol:73]: The var internalCollateralToUse isn't used to after https://github.com/horse-link/contracts.horse.link/pull/99/files#diff-f5900bd5f7e2d9283f2b98c019a41f01b7792490e99c877ac08b20c7106e4acfR78

Comment (bitcoinbrisbane) [reply contracts/MarketCollateralised.sol:73]: > I don't follow

Lets move it to inside the "if" condition on LN78

File: contracts/MarketCollateralised.sol [hunk 7]
```diff
@@ -68,44 +68,60 @@ abstract contract MarketCollateralised is Market {
...
+				amountToBorrow
+			);
+
+			// Add the amount we borrowed to the total collateral				
+			emit Borrowed(index, amountToBorrow);
+			_totalCollateral += amountToBorrow;
+		}
+
+		// Store the amount of exposure that this bet added. This will be deducted from the total exposure when the bet is paid out
+		_betExposure[_bets.length] = amountToObtain;
```
Comment (brettvniekerk) [inline contracts/MarketCollateralised.sol:113]: quick q on why we map the `_betExposure` by the total bets and not the bet `index` ?

**Review by brettvniekerk — APPROVED** [review]

File: contracts/MarketCollateralised.sol (hunk 7 above)
Comment (bazmatic) [reply contracts/MarketCollateralised.sol:113]: Because the bet hasn't been added the the array yet

Comment (bazmatic) [reply contracts/MarketCollateralised.sol]: I see, we actually do have the `index` there! I'll change it.

**Review by brettvniekerk — APPROVED** [review]

**Review by bitcoinbrisbane — APPROVED** [review]

**Pull request merged**